src/evm/protocol/uniswap_v4/state.rs

```diff
@@ -43,6 +67,200 @@ impl UniswapV4State {
+    ) -> Result<SwapResults, SimulationError> {
+        if self.liquidity == 0 {
+            return Err(SimulationError::RecoverableError("No liquidity".to_string()));
+        }
+        let price_limit = if let Some(limit) = sqrt_price_limit {
+            limit
+        } else if zero_for_one {
+            safe_add_u256(MIN_SQRT_RATIO, U256::from(1u64))?
+        } else {
+            safe_sub_u256(MAX_SQRT_RATIO, U256::from(1u64))?
+        };
+
+        if zero_for_one {
+            assert!(price_limit > MIN_SQRT_RATIO);
+            assert!(price_limit < self.sqrt_price);
+        } else {
+            assert!(price_limit < MAX_SQRT_RATIO);
+            assert!(price_limit > self.sqrt_price);
+        }
+
+        let exact_input = amount_specified > I256::from_raw(U256::from(0u64));
+
+        let mut state = SwapState {
+            amount_remaining: amount_specified,
+            amount_calculated: I256::from_raw(U256::from(0u64)),
+            sqrt_price: self.sqrt_price,
+            tick: self.tick,
+            liquidity: self.liquidity,
+        };
+        let mut gas_used = U256::from(130_000);
+
+        while state.amount_remaining != I256::from_raw(U256::from(0u64)) &&
+            state.sqrt_price != price_limit
+        {
+            let (mut next_tick, initialized) = match self
+                .ticks
+                .next_initialized_tick_within_one_word(state.tick, zero_for_one)
+            {
+                Ok((tick, init)) => (tick, init),
+                Err(tick_err) => match tick_err.kind {
+                    TickListErrorKind::TicksExeeded => {
+                        let mut new_state = self.clone();
+                        new_state.liquidity = state.liquidity;
+                        new_state.tick = state.tick;
+                        new_state.sqrt_price = state.sqrt_price;
+                        return Err(SimulationError::InvalidInput(
+                            "Ticks exceeded".into(),
+                            Some(GetAmountOutResult::new(
+                                u256_to_biguint(state.amount_calculated.abs().into_raw()),
+                                u256_to_biguint(gas_used),
+                                Box::new(new_state),
+                            )),
+                        ));
+                    }
+                    _ => return Err(SimulationError::FatalError("Unknown error".to_string())),
+                },
+            };
+
+            next_tick = next_tick.clamp(MIN_TICK, MAX_TICK);
+
+            let sqrt_price_next = get_sqrt_ratio_at_tick(next_tick)?;
+            let (sqrt_price, amount_in, amount_out, fee_amount) = swap_math::compute_swap_step(
+                state.sqrt_price,
+                UniswapV4State::get_sqrt_ratio_target(sqrt_price_next, price_limit, zero_for_one),
+                state.liquidity,
+                state.amount_remaining,
+                self.fees
+                    .calculate_swap_fees_pips(zero_for_one),
```

Comment:
So basically this line is the only difference with usv3 right now? 👀

Reply:
It's not even really a difference because it's there in V3, we just don't have it implemented because they never decided to enable protocol fees.

I'd say the biggest difference I've seen do far is that the `fees` and `tick_spacing` can be anything and there is no map `fee => tick_spacing` anymore. I'm looking forward to test all this with actual on chain state to confirm though